mojo/proposals/drain-collection-support.md

```diff
@@ -0,0 +1,452 @@
+zip them together into a map-like object (Dict, Tree, etc.).
+
+### How do other languages handle this?
+
+Rust
+
+> TODO: Would like input from people with Rust heavy backgrounds to clarify this
+> is how Rust drain actually works, particularly "when do the source collections
+> get cleared"?
+
+Rust has a struct called
+[`vec/Drain`][2] and
+[`hashmap/Drain`][3]
+that is used by collections like `Vec` and `HashMap` to drain the collection.
+
+[Vec.drain](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain)
+
+```rust
+let mut v = vec![1, 2, 3];
+let u: Vec<_> = v.drain(1..).collect();
+assert_eq!(v, &[1]);
+assert_eq!(u, &[2, 3]);
+
+// A full range clears the vector, like `clear()` does
+v.drain(..);
+assert_eq!(v, &[]);
+```
+
+[HashMap.drain](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.drain)
+
+```rust
+use std::collections::HashMap;
+
+let mut a = HashMap::new();
+a.insert(1, "a");
+a.insert(2, "b");
+
+for (k, v) in a.drain().take(1) {
+    assert!(k == 1 || k == 2);
+    assert!(v == "a" || v == "b");
+}
+
+assert!(a.is_empty());
+```
+
+For sequential collections like Vec:
+
+- `drain(..)` clears the full collection either on full iteration or when the
+iterator is dropped (destroyed).
+- `drain(start..end)` drains a range of elements either on full iteration or when
+the iterator is dropped (destroyed).
+
+For hashable collections like HashMap:
+
+- `drain().take(n)` drains the first `n` elements. *Clears the entire collection
+on destruction.*
+- `drain()` drains the full collection either on full iteration or when the
+iterator is dropped (destroyed).
+
+Note that the `Drain` struct has [`Drain.keep_rest`][4]
+that avoids clearing unyielded elements when Drain is dropped.
+
+## Proposal
+
+The stdlib team has proposed to move forward with [Rust style draining of
+collections][5].
+
+Scope:
+
+This proposal can expand to support additional
+collections beyond `List`, `Set`, and `Dict`, however for brevity we will focus
+on the basic collections for now. It also goes through compatibility with iter
+tools like `zip`, however will not go into compatibility of all the iter tool
+functions that could be added in the future that Python supports.
+
+Goals:
+
+- `take_items` implementation for `Set`, `List`, and `Dict` initially that aligns
+with Rust's `Drain` struct that does not copy elements.
+- Compatibility with iter tools like `zip`.
+
+Non-Goals:
+
+- Block on move-only tuple unpacking.
+
+### User Experience Cases
+
+In all cases, `ref` must be used to get mutable access to the item. `for` loops
+by default iterate over immutable references (even if the source collection is
+mutable).
+
+In all cases the call to `take_items` takes mutable reference of the underlying
+collection, and clears the source collection immediately on destruction of the
+iterator. Cases such as `take_items(Slice(..))` takes mutable reference of the
```

Comment:
FWIW: I wonder if taking a `range` is better? Since the only nice part about `slice` is the ability to use the `[:]` syntax with it. But the Mojo parser only supports the `[:]` syntax in combination with square brackets.

Reply:
So you'll be able to write `list.drain(range(1, 10))` which is much cleaner to read. Also we probably should not support backwards ranges? That feels like it will drive the complexity of the implementation to be too large.